pkg/driver/clients/grpc_client.go

```diff
@@ -87,38 +89,63 @@ func (i *grpcInvoker) Invoke(function *common.Function, runtimeSpec *common.Runt
-	defer cancelExecution()
-
-	response, err := grpcClient.Execute(executionCxt, &proto.FaasRequest{
-		Message:           "nothing",
-		RuntimeInMilliSec: uint32(runtimeSpec.Runtime),
-		MemoryInMebiBytes: uint32(runtimeSpec.Memory),
-	})
-
-	if err != nil {
-		logrus.Debugf("gRPC timeout exceeded for function %s - %s", function.Name, err)
 
+	defer cancelExecution()
+	if !i.cfg.VSwarm {
+		grpcClient := protoExec.NewExecutorClient(conn)
+		response, err := grpcClient.Execute(executionCxt, &protoExec.FaasRequest{
+			Message:           "nothing",
+			RuntimeInMilliSec: uint32(runtimeSpec.Runtime),
+			MemoryInMebiBytes: uint32(runtimeSpec.Memory),
+		})
+		if err != nil {
+			logrus.Debugf("gRPC timeout exceeded for function %s - %s", function.Name, err)
+
+			record.ResponseTime = time.Since(start).Microseconds()
+			record.FunctionTimeout = true
+
+			return false, record
+		}
+		record.Instance = extractInstanceName(response.GetMessage())
 		record.ResponseTime = time.Since(start).Microseconds()
-		record.ConnectionTimeout = true // WithBlock deprecated in new gRPC interface
-		record.FunctionTimeout = true
-
-		return false, record
-	}
-
-	record.Instance = extractInstanceName(response.GetMessage())
-	record.ResponseTime = time.Since(start).Microseconds()
-	record.ActualDuration = response.DurationInMicroSec
-
-	if strings.HasPrefix(response.GetMessage(), "FAILURE - mem_alloc") {
-		record.MemoryAllocationTimeout = true
+		record.ActualDuration = response.DurationInMicroSec
+
+		if strings.HasPrefix(response.GetMessage(), "FAILURE - mem_alloc") {
+			record.MemoryAllocationTimeout = true
+		} else {
+			record.ActualMemoryUsage = common.Kib2Mib(response.MemoryUsageInKb)
+		}
+		logrus.Tracef("(Replied)\t %s: %s, %.2f[ms], %d[MiB]", function.Name, response.Message,
+			float64(response.DurationInMicroSec)/1e3, common.Kib2Mib(response.MemoryUsageInKb))
 	} else {
-		record.ActualMemoryUsage = common.Kib2Mib(response.MemoryUsageInKb)
+		grpcClient := proto.NewGreeterClient(conn)
+		response, err := grpcClient.SayHello(executionCxt, &proto.HelloRequest{
+			Name: "Invoke Relay",
+			VHiveMetadata: MakeVHiveMetadata(
+				uuid.New().String(),
+				uuid.New().String(),
+				time.Now().UTC(),
+			),
+		})
+		if err != nil {
+			logrus.Debugf("gRPC timeout exceeded for function %s - %s", function.Name, err)
+
+			record.ResponseTime = time.Since(start).Microseconds()
+			record.FunctionTimeout = true
+
+			return false, record
+		}
+		record.ResponseTime = time.Since(start).Microseconds()
+		record.ActualDuration = uint32(record.ResponseTime)
+		record.Instance = extractInstanceName(response.GetMessage())
+		if strings.HasPrefix(response.GetMessage(), "FAILURE - mem_alloc") {
+			record.MemoryAllocationTimeout = true
+		} else {
+			record.ActualMemoryUsage = common.Kib2Mib(0) //common.Kib2Mib(response.MemoryUsageInKb)
+		}
```

Comment:
Extract this code into a new vSwarmGrpcInvoker instead of making a special case for the grpcInvoker